sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp

```diff
@@ -0,0 +1,31 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+template <usm::alloc alloc_kind> void test(queue &q) {
+  sycl::usm_allocator<int, alloc_kind> ua(q);
+  int *p = ua.allocate(0);
+
+  assert(!p && "Our implementation of usm_allocator is expected to return a "
+               "null pointer when allocation "
+               "size is zero.");
```

Comment:
I think this is fine, but another option is to simply remove the assert.  From an API perspective, the important thing is that you can call allocate with a zero size and that you can then call deallocate on the returned pointer, all without any exception being raised.  It's not important whether or not the returned pointer is null.